### PR TITLE
Cache liveblog archive pages for longer

### PR DESCRIPTION
**Pull request by johnduffell**
Liveblogs have a cache time of 5 seconds as they update frequently.  This also applied to archive pages, which won't update as much.  Since frontend fetches the entire JSON from capi on every render, this was causing a high load for very long live blogs (e.g. the 17 page long one last week)

This PR implements a simple solution to reduce the load by increasing the cache time for archive pages to be the same as for an equivalent article (probably 300 seconds for a liveblog that's being updated regularly)

After that most of the load would come from the first page.  A separate piece of work is under way to retrieve only the last n blocks from capi - this will help with live updates and loading the first page.

I had to refactor things a bit because we were assuming one resource in capi mapped to one http response from frontend.  This meant the cache time was baked into the model we populate from capi data.  I have integrated the live blog page model more closely into the capi model and separated out the cache time a little more clearly.

@rich-nguyen @cb372

### CODE REVIEW

File: common/test/recorder/HttpRecorder.scala [hunk 1]
```diff
@@ -19,8 +19,6 @@ trait HttpRecorder[A] extends ExecutionContexts {
   final def load(url: String, headers: Map[String, String] = Map.empty)(fetch: => Future[A]):Future[A] = {
 
     val fileName = name(url, headers)
-    println(url)
-    println(fileName)
```
Comment (johnduffell) [inline common/test/recorder/HttpRecorder.scala]: @TBonnin :stuck_out_tongue_closed_eyes:

Comment (TBonnin) [inline common/test/recorder/HttpRecorder.scala]: Damn!!!! Thanks for fixing

File: article/app/controllers/ArticleController.scala [hunk 2]
```diff
@@ -192,18 +172,66 @@ object ArticleController extends Controller with RendersItemResponse with Loggin
...
+      maybeRequiredBlockId.map(blockId => block => blockId == block.id),
+      _.id
+    )
+    liveBlogPageModel match {
+      case Some(pageModel) =>
+
+        val cacheTime =
+          if (!pageModel.currentPage.isArchivePage && liveBlog.fields.isLive) liveBlog.metadata.cacheTime
+          else {
+            if (LongCacheSwitch.isSwitchedOn) {
```
Comment (johnduffell) [inline article/app/controllers/ArticleController.scala]: not completely happy to replicate all this logic here as well

File: common/app/model/Formats.scala [hunk 3]
```diff
@@ -75,7 +75,6 @@ object MetaDataFormat {
     contentType: String,
     hasHeader: Boolean,
     schemaType: Option[String],
-    cacheSeconds: Int,
```
Comment (johnduffell) [inline common/app/model/Formats.scala]: not sure if it's worth storing the cache time in the pressed json, unless I missed something?

**Comment by johnduffell** [comment]
I've had another go @rich-nguyen - see what you think?

**Comment by johnduffell** [comment]
conflicts already! :cry: come on @rich-nguyen :stuck_out_tongue:

File: article/app/controllers/ArticleController.scala [hunk 4]
```diff
@@ -192,18 +172,66 @@ object ArticleController extends Controller with RendersItemResponse with Loggin
...
+        case (minute: Article, None) if minute.isUSMinute =>
+          Left(MinutePage(minute, RelatedContent(minute, response)))
+        case (liveBlog: Article, None) if liveBlog.isLiveBlog =>
+          createLiveBlogModel(liveBlog, response, None)
+        case (liveBlog: Article, Some(Some(requiredBlockId))) if liveBlog.isLiveBlog =>
+          createLiveBlogModel(liveBlog, response, Some(requiredBlockId))
+        case (article: Article, None) =>
+          Left(ArticlePage(article, RelatedContent(article, response)))
+        case _ =>
+          Right(NotFound)
```
Comment (cb372) [inline article/app/controllers/ArticleController.scala]: Left for success, right for failure? Controversial!

File: article/app/controllers/ArticleController.scala [hunk 5]
```diff
@@ -192,18 +172,66 @@ object ArticleController extends Controller with RendersItemResponse with Loggin
...
-      case article: Article => ArticlePage(article, RelatedContent(article, response))
+    val page = pageParam.map(ParseBlockId.apply)
+    val supportedContentResult = ModelOrResult(supportedContent, response)
+    val content: Either[PageWithStoryPackage, Result] = supportedContentResult.left.flatMap { content =>
+      (content, page) match {
+        case (minute: Article, None) if minute.isUSMinute =>
+          Left(MinutePage(minute, RelatedContent(minute, response)))
+        case (liveBlog: Article, None) if liveBlog.isLiveBlog =>
+          createLiveBlogModel(liveBlog, response, None)
+        case (liveBlog: Article, Some(Some(requiredBlockId))) if liveBlog.isLiveBlog =>
```
Comment (cb372) [inline article/app/controllers/ArticleController.scala]: I'm sure there's a good reason for it, but the nested `Option` here smells strange to me

File: article/test/liveblog/PageParserTest.scala [hunk 6]
```diff
@@ -1,18 +1,18 @@
 package liveblog
 
-import controllers.PageParser
+import controllers.ParseBlockId
 import org.scalatest.{FlatSpec, Matchers}
 
 class PageParserTest extends FlatSpec with Matchers {
```
Comment (cb372) [inline article/test/liveblog/PageParserTest.scala]: Could rename this test to `ParseBlockIdTest`?

**Comment by cb372** [comment]
Looks good to my unqualified eye

File: common/app/model/Formats.scala [hunk 7]
```diff
@@ -110,7 +109,7 @@ object MetaDataFormat {
         part2.contentType,
         part2.hasHeader,
         part2.schemaType,
-        part2.cacheSeconds,
```
Comment (rich-nguyen) [inline common/app/model/Formats.scala]: The only reason to keep the cache seconds in the pressed json is to keep serialisation simple. So we wouldn't need too many rules describing which fields get 'slimmed', which ones become default, etc. Here, the deserialisation is a bit more complicated.

File: article/app/controllers/ArticleController.scala (hunk 4 above)
Comment (johnduffell) [inline article/app/controllers/ArticleController.scala]: maybe you could class it as a success to come up with a Result without having to do any more processing :stuck_out_tongue_closed_eyes:

File: article/app/controllers/ArticleController.scala (hunk 5 above)
Comment (johnduffell) [inline article/app/controllers/ArticleController.scala]: the outer one is whether the user specified a block id, and the inner one was whether the one they specified was actually valid.  If they specified one which is invalid, it's a 404.  Maybe I could do some matching onto case classes or something clever, or at least add a comment though.

File: article/test/liveblog/PageParserTest.scala (hunk 6 above)
Comment (johnduffell) [inline article/test/liveblog/PageParserTest.scala]: good point

**Comment by rich-nguyen** [comment]
to recap:
- the `LiveBlogPageModel` could be simpler without the typeclass, given it always takes `BodyBlock`
- the format should match the original model as much as possible, including `cacheTime`
- feel free to name things more sensibly now, eg `LiveBlogPage` -> `LiveBlog`

Great :+1:

**Pull request merged**